### ISSUE TITLE
Incorrect marshalling of CXUnsavedFile unsaved_files arguments to clang_parseTranslationUnit and friends

### ISSUE THREAD

**Issue by Shauren** [microsoft/ClangSharp#12]
Currently the parameters are generated as `out CXUnsavedFile` however according to documentation http://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html it is actually an array of files (and the count is passed as separate argument)
This makes it impossible to pass more then one CXUnsavedFile

The correct form would be `[MarshalAs(UnmanagedType.LPArray)] CXUnsavedFile[] @unsaved_files`

This bug probably applies to all functions that take an array argument as pair of pointer to first element and count.

**Comment by mjsabby** [comment]
Thanks for the report. This is much like some of the other issues where marshalling is awkward because of the auto-generated code.

I'm going to be solving this in a more holistic way, by keeping the higher-level features like array marshalling separate from the auto-generated code. The code will be backwards compatible, but the autogenerator will need a way of ignoring certain methods.

TL;DR Coming soon.

**Comment by danwalmsley** [comment]
I'm also looking forward to this fix :)

**Comment by mjsabby** [comment]
I was hoping to make my changes in a couple of days hence my previous comment, but since I've not managed to finish it, I'll put this out there to unblock both of you and anyone else looking.

The idea is to make your array as usual, but when calling the ClangSharp method of interest, you pass in a reference to the first element, like ``arr[0]``, but pass in the correct length ``arr.Length``

This is the awkward part, i.e. calling it, but the marshalling itself is correct.

```csharp
string filename = ...
CXTranslationUnit tu = ...
var unsaved_files = new CXUnsavedFile[2];
clang.codeCompleteAt(tu, filename, 0, 0, out unsaved_files[0], (uint)unsaved_files.Length, 0);
```

Let me know if this doesn't unblock you.

**Comment by Shauren** [comment]
That does not work for me, trying to use it like that ends with a crash in libclang
> libclang: crash detected during parsing: {
  'source_filename' : 'DB2Structure.h'
  'command_line_args' : ['-x', 'c++'],
  'unsaved_files' : [('(null)', '...', 140608701536024)],
  'options' : 64,
}

https://github.com/Shauren/WowClientDB2MySQLTableGenerator it the project I am using ClangSharp in, a tiny console utility producing no output to stdout, requiring this file to be present in exe directory https://gist.github.com/Shauren/055a58e632873af65742cee228792e35, creating 3 output files (.sql, .cpp and .h)
To reproduce it in that project simply change the call in this line https://github.com/Shauren/WowClientDB2MySQLTableGenerator/blob/master/HeaderParser.cs#L66 to original version from clang namespace

I worked around this by simply declaring the DllImport with proper marshalling in my own namespace as seen here
https://github.com/Shauren/WowClientDB2MySQLTableGenerator/blob/master/HeaderParser.cs#L13

**Comment by mjsabby** [comment]
Alright, I'll take a look. Do I also need Common.h, DBCEnums.h and Utils.h to clone and make your tool work?

**Comment by Shauren** [comment]
No, all includes are dynamically removed and put into the CXUnsavedFile
https://github.com/Shauren/WowClientDB2MySQLTableGenerator/blob/master/HeaderParser.cs#L53

**Comment by mjsabby** [comment]
Looks like it has to ``ref`` and not ``out`` for my trick to work. Since that requires a ClangSharp change anyway, I'll be looking at out parameter arguments to see what's broken like this.

**Comment by MxKy** [comment]
Is a fix still incoming for this bug?

As I understand it, it is currently impossible to e.g. use `clang.parseTranslationUnit()` with even a single `CXUnsavedFile`.

Is there currently any workaround that avoids a crash like this?

> libclang: crash detected during parsing: {
>   'source_filename' : 'test.c'
>   'command_line_args' : ['clang'],
>   'unsaved_files' : [('(null)', '...', 0)],
>   'options' : 0,
> }

**Comment by mjsabby** [comment]
@MxKy yup, I'll do this soon.

**Comment by jairbubbles** [comment]
@mjsabby  I guess I'm in the same situation than @MxKy. I can't use `clang.parseTranslationUnit()` at all. Whatever the version (3.6, 3.8 or even with upstream version) I get a `crash detected during parsing:`.

I feel like it's another issue as seeing you're using it in `ClangSharpPInvokeGenerator` well I guess it works.

**Comment by ForeverZer0** [comment]
This bug remains, and there are a plethora of examples an incorrect use of `out` all through the project.  I understand for code generation, there is always going to be some things that need fixed up manually, and edge-cases that no amount of coding is going to solve elegantly, but the majority of the examples within the main project could be solved by simply changing from an `out TYPE` to `IntPtr`.

**Comment by tannergooding** [comment]
This should be resolved now. `parseTranslationUnit` now takes a `CXUnsavedFile[]` and the generator was updated to automatically generate the correct code here.

Would be helpful if someone could validate it is fixed, since this has been non-trivial in the past. Would also be helpful to see if this is resolved without the https://github.com/Microsoft/ClangSharp/blob/master/ClangSharp/Generated.Custom.cs#L42 marshaling wrapper

**Comment by tannergooding** [comment]
The testing framework is now utilizing `CXUnsaveFile` for its validation process so I'm fairly confident this is working as intended now.

I'm going to close this, but feel free to reopen if you can reproduce the issue.
